include/albatross/src/stats/incomplete_gamma.hpp

```diff
@@ -0,0 +1,178 @@
+  return (ub - lb) * 0.5 * gauss_legendre_50_weights[counter];
+}
+
+inline double incomplete_gamma_quadrature_fn(double x, double a,
+                                             double large_term) {
+  return exp(-x + (a - 1.) * log(x) - large_term);
+}
+
+inline double incomplete_gamma_quadrature_recursive(double lb, double ub,
+                                                    double a, double lg_term,
+                                                    std::size_t counter) {
+  if (counter < gauss_legendre_50_weights.size() - 1) {
+    return incomplete_gamma_quadrature_fn(
+               incomplete_gamma_quadrature_inp_vals(lb, ub, counter), a,
+               lg_term) *
+               incomplete_gamma_quadrature_weight_vals(lb, ub, counter) +
+           incomplete_gamma_quadrature_recursive(lb, ub, a, lg_term,
+                                                 counter + 1);
+  } else {
+    return incomplete_gamma_quadrature_fn(
+               incomplete_gamma_quadrature_inp_vals(lb, ub, counter), a,
+               lg_term) *
+           incomplete_gamma_quadrature_weight_vals(lb, ub, counter);
+  }
+}
+
+inline double incomplete_gamma_quadrature_lb(double a, double z) {
+  if (a > 1000) {
+    return std::max(0., std::min(z, a) - 11 * sqrt(a));
+  } else if (a > 800) {
+    return std::max(0., std::min(z, a) - 11 * sqrt(a));
+  } else if (a > 500) {
+    return std::max(0., std::min(z, a) - 10 * sqrt(a));
+  } else if (a > 300) {
+    return std::max(0., std::min(z, a) - 10 * sqrt(a));
+  } else if (a > 100) {
+    return std::max(0., std::min(z, a) - 9 * sqrt(a));
+  } else if (a > 90) {
+    return std::max(0., std::min(z, a) - 9 * sqrt(a));
+  } else if (a > 70) {
+    return std::max(0., std::min(z, a) - 8 * sqrt(a));
+  } else if (a > 50) {
+    return std::max(0., std::min(z, a) - 7 * sqrt(a));
+  } else if (a > 40) {
+    return std::max(0., std::min(z, a) - 6 * sqrt(a));
+  } else if (a > 30) {
+    return std::max(0., std::min(z, a) - 5 * sqrt(a));
+  }
+  return std::max(0., std::min(z, a) - 4 * sqrt(a));
+}
+
+inline double incomplete_gamma_quadrature_ub(double a, double z) {
+  if (a > 1000) {
+    return std::min(z, a + 10 * sqrt(a));
+  } else if (a > 800) {
+    return std::min(z, a + 10 * sqrt(a));
+  } else if (a > 500) {
+    return std::min(z, a + 9 * sqrt(a));
+  } else if (a > 300) {
+    std::min(z, a + 9 * sqrt(a));
+  } else if (a > 100) {
+    return std::min(z, a + 8 * sqrt(a));
+  } else if (a > 90) {
+    return std::min(z, a + 8 * sqrt(a));
+  } else if (a > 70) {
+    return std::min(z, a + 7 * sqrt(a));
+  } else if (a > 50) {
+    return std::min(z, a + 6 * sqrt(a));
+  } else if (a > 40) {
+    return std::min(z, a + 5 * sqrt(a));
+  }
+  return std::min(z, a + 4 * sqrt(a));
+}
+
+double incomplete_gamma_quadrature(double a, double z) {
+  return incomplete_gamma_quadrature_recursive(
+      incomplete_gamma_quadrature_lb(a, z),
+      incomplete_gamma_quadrature_ub(a, z), a, lgamma(a), 0);
```

Comment:
The upper and lower bounds quadrature functions seem heavily related. I might consider merging them into a single function that returns a tuple/pair/struct. I think that'll make the conditionals more intuitive if they are zippered together.